src/reject.rs

```diff
@@ -410,13 +424,29 @@ impl Rejections {
                 | Known::MissingExtension(_)
                 | Known::BodyConsumedMultipleTimes(_) => StatusCode::INTERNAL_SERVER_ERROR,
             },
+            Rejections::Unauthorized(..) => StatusCode::UNAUTHORIZED,
             Rejections::Custom(..) => StatusCode::INTERNAL_SERVER_ERROR,
             Rejections::Combined(ref a, ref b) => preferred(a, b).status(),
         }
     }
 
     fn into_response(&self) -> crate::reply::Response {
         match *self {
+            Rejections::Unauthorized(ref e) => {
+                let body = format!("Please Authorize for: {:?}", e);
+                let mut res = http::Response::new(Body::from(body));
+                *res.status_mut() = self.status();
+                res.headers_mut().insert(
+                    CONTENT_TYPE,
+                    HeaderValue::from_static("text/plain; charset=utf-8"),
+                );
+                let realm = e.downcast_ref::<String>().expect("realm must be string");
+                res.headers_mut().insert(
+                    "www-authenticate",
+                    HeaderValue::from_str(format!("Basic realm=\"{}\"", realm).as_str()).unwrap(),
```

Comment:
The unauthorized response should be able to handle both `Basic` and `Bearer` replies at least. [The full list of valid values here is 10 entries long](http://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml), so some flexibility on the `scheme` would be nice.

Also according to [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate) the `realm` _could_ be omitted, though I am not sure it is wise to allow that

Reply:
Given that what I've coded so far feels kinda hacky, supporting multiple schemes calls for an even cleaner implementation. Meaning for example coming up with a new rejection type as mentioned in my description of the PR? I'll try to add `Bearer` but I am not sure this is really readable code after without fundamental changes.

Reply:
Updating the [error type](https://github.com/seanmonstar/warp/pull/751/files/7b357e450c292dc252d40325d2aa82c56994a3e9#diff-3b90cd596f94fb9464f356d926c42bb98767b769f938613f9f7e971d3e328c3cR548) to not be a `unit_error` should simplify the realm and scheme extraction from the user's configuration.

Reply:
On issue #62 there are a few files still available from the previous PR around an auth filter. I've updated their approach to the current master [here](https://github.com/FreeMasen/warp/blob/auth_filter/src/reject.rs#L111) which might be worth looking at.

Reply:
thx for the feedback, I'll hack away :)